### PR TITLE
[fix][tinker] Preserve gradient accumulation across multiple forward_backward calls prior to optim_step in Megatron

### PR DESCRIPTION
**Pull request by yapdianang**
## What changed

- Megatron keeps gradients from every `forward_backward` call until `optim_step` updates the model.
- Tests cover sending the same training data as one call or several calls.

## Problem

Tinker's `TrainingClient` can divide one `forward_backward(data)` call into several requests when the configured size limits are reached. Each request is sent to `/forward_backward` separately: [Tinker source](https://github.com/thinking-machines-lab/tinker/blob/d595dc2ba62a1494aa0f452fe1d0dbb7b321b870/src/tinker/lib/public_interfaces/training_client.py#L180-L210).

The Megatron worker calls `zero_grad_buffer()` at the start of every `forward_backward` call: [SkyRL source](https://github.com/NovaSky-AI/SkyRL/blob/fe1ad6c154f6c40bfcfb3c514f5afe3b1dbed5a4/skyrl/backends/skyrl_train/workers/megatron/megatron_worker.py#L1130-L1133).

These two behaviors work separately but not together. If one model update arrives in several requests, each request erases the gradients from the requests before it. `optim_step` then uses only the last request.

## Test

The added Megatron GPU test creates two fresh LoRA models with the same starting weights and training data.

The first model receives:

```python
forward_backward(first)
forward_backward(second)
forward_backward(final_zero)
optim_step()
```

The second receives:

```python
forward_backward([first, second, final_zero])
optim_step()
```

`final_zero` has zero training weight. Before this change, it clears the first model's earlier gradients and adds none of its own. The two models therefore receive different updates. After this change, both models should receive the same update.

See [`test_forward_backward_accumulates_across_requests`](https://github.com/NovaSky-AI/SkyRL/blob/33cb38bd3b0960da5fec729d74b997511fed39b2/tests/tinker/skyrl_train/test_multi_lora_megatron.py#L278-L304).

## Why it matters

For Tinker training through Megatron, a large batch can report losses for all examples while only the last request changes the model. Training continues without an error, so the problem is easy to miss. Batch size and example order can change the result.

This does not affect a model update that reaches Megatron in one `forward_backward` call.

## Checks

- `ruff check`
- `tests/tinker/test_engine.py`: 14 passed; two JAX tests could not import on this host
- Code quality, CPU training, CPU, and gym GitHub checks passed

---

> [!NOTE]
> **High Risk**
> Changes core Megatron training gradient lifecycle (sync timing, buffer clearing) on the Tinker path; incorrect behavior would silently wrong-foot large-batch training when DP > 1, though new GPU tests target the failure mode.
> 
> **Overview**
> Fixes **Megatron Tinker training** when one logical update is split across several `forward_backward` HTTP requests. Gradients from every call now accumulate until `optim_step`, instead of being cleared or corrupted mid-window.
> 
> **Megatron worker / model wrapper:** `forward_backward` no longer calls `zero_grad_buffer()` at entry. Megatron’s end-of-schedule `finalize_model_grads` is **deferred** via `_defer_finalize_model_grads` and run **once** in `optim_step` through `run_pending_grad_sync`, because DP reduce-scatter on `grad_data` is **not idempotent** when repeated across multiple calls in one accumulation window. After the optimizer step, `zero_grad_buffer()` clears DDP `grad_data` for the next window. Ranks that see no microbatches still join the deferred collective to avoid hangs.
> 
> **Docs & multi-LoRA:** Tinker architecture docs describe cross-request gradient accumulation; `AdapterStore` comments clarify that per-adapter CPU grad mirrors support interleaved tenants.
> 
> **Tests:** Multi-LoRA Megatron GPU tests now use **DP=2** (≥3 GPUs) and add cases comparing split vs combined `forward_backward`, all DP ranks fed, and zero-weight chunks as no-ops.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 4db102ca6d33f6dfea2bbd33eebab8fbf72b346c. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by yapdianang** [comment]
Additional context for review:

1. Tinker's `TrainingClient` divides `forward_backward(data)` when the configured request size is reached: [request splitting](https://github.com/thinking-machines-lab/tinker/blob/d595dc2ba62a1494aa0f452fe1d0dbb7b321b870/src/tinker/lib/public_interfaces/training_client.py#L180-L210).
2. It sends each resulting request through `client.training.forward_backward(...)`: [request submission](https://github.com/thinking-machines-lab/tinker/blob/d595dc2ba62a1494aa0f452fe1d0dbb7b321b870/src/tinker/lib/public_interfaces/training_client.py#L287-L325).
3. The Megatron worker calls `zero_grad_buffer()` at the start of every `forward_backward`: [SkyRL source](https://github.com/NovaSky-AI/SkyRL/blob/fe1ad6c154f6c40bfcfb3c514f5afe3b1dbed5a4/skyrl/backends/skyrl_train/workers/megatron/megatron_worker.py#L1130-L1133).

These two behaviors work separately but not together. If one model update is sent in several requests, each request erases the gradients from the requests before it. Only the last request reaches `optim_step`.

This PR adds a direct test using two fresh LoRA models. One receives three `forward_backward` calls before `optim_step`; the other receives the same data in one call. The last datum has zero training weight. Before the fix, that final call clears the earlier gradients and leaves a zero update. After the fix, both models should receive the same update: [test source](https://github.com/NovaSky-AI/SkyRL/blob/33cb38bd3b0960da5fec729d74b997511fed39b2/tests/tinker/skyrl_train/test_multi_lora_megatron.py#L278-L304).

**Comment by erictang000** [comment]
@cursor review

File: skyrl/backends/skyrl_train/workers/megatron/megatron_worker.py [hunk 1]
```diff
@@ -1306,14 +1305,33 @@ def optim_step(self) -> Optional[float]:
...
         grad_norm = self.strategy.optimizer_step(self.optimizer, self.model, self.scheduler, name="actor")
 
+        # Clear the DDP grad buffers for the next window. `optimizer.zero_grad()` inside
+        # `optimizer_step` only drops `param.grad` / the fp32 main-param grads -- the
+        # `grad_data` buffer that `param.main_grad` views is untouched, so without this
+        # the gradients just applied would be accumulated into again by the next window.
+        # Also re-arms Megatron's per-iteration bookkeeping (bucket-group grad-ready
+        # counters, `grad_added_to_main_grad`).
+        for chunk in self.actor_module:
+            chunk.zero_grad_buffer()
```
[the remark this answers is not in the document]
Comment (erictang000) [reply skyrl/backends/skyrl_train/workers/megatron/megatron_worker.py:1335]: i think this is fine, let's assume that all optim_steps are successful.

File: skyrl/backends/skyrl_train/workers/megatron/megatron_model_wrapper.py [hunk 2]
```diff
@@ -209,16 +209,58 @@ def __init__(
...
+
+        ``num_tokens`` is only non-None under ``calculate_per_token_loss`` (never set by
+        SkyRL, whose loss scaling assumes the per-microbatch path); accumulate it across
+        calls so the deferred sync divides by the whole window's token count if it ever is.
+        """
+        del model, kwargs  # replayed against self.actor_module with default process groups
+        pending = self._pending_grad_sync
+        if pending is not None and pending["num_tokens"] is not None and num_tokens is not None:
+            num_tokens = pending["num_tokens"] + num_tokens
+        self._pending_grad_sync = {"num_tokens": num_tokens}
```
[the remark this answers is not in the document]
Comment (erictang000) [reply skyrl/backends/skyrl_train/workers/megatron/megatron_model_wrapper.py:251]: this seems unlikely

**Review by pcmoritz — APPROVED** [review]
Thanks a lot for updating this :)

**Review by erictang000 — APPROVED** [review]
thanks for reporting this @yapdianang
